File: src/dots/core/system.py
import platform
import sys
from pathlib import Path
from typing import Optional

def detect_os() -> str:
    """
    Detect the current operating system.
    Returns: 'linux', 'mac', 'windows', or 'unknown'
    """
    system = platform.system().lower()
    if system == "linux":
        return "linux"
    elif system == "darwin":
        return "mac"
    elif system == "windows" or "mingw" in system or "cygwin" in system:
        return "windows"
    return "unknown"

def get_home_dir() -> Path:
    """
    Get the user's home directory.
    """
    return Path.home()

def is_safe_path(path: Path) -> bool:
    """
    Check if a path is safe to modify (inside HOME).
    """
    try:
        home = get_home_dir().resolve()
        target = path.resolve()
        
        # Must be inside home
        if not str(target).startswith(str(home)):
            return False
            
        # Must not be home itself
        if target == home:
            return False
            
        return True
    except Exception:
        return False

File: internal/system/system.go
package system

import (
	"os"
	"os/user"
	"path/filepath"
	"runtime"
	"strings"
)

// DetectOS returns the current operating system: "linux", "mac", "windows", or "unknown".
func DetectOS() string {
	switch runtime.GOOS {
	case "linux":
		return "linux"
	case "darwin":
		return "mac"
	case "windows":
		return "windows"
	default:
		return "unknown"
	}
}

// IsSafePath checks if a path is safe to modify (inside HOME and not HOME itself).
func IsSafePath(path string) bool {
	home := getHomeDir()
	if home == "" {
		return false
	}

	absPath, err := filepath.Abs(path)
	if err != nil {
		return false
	}

	absHome, err := filepath.Abs(home)
	if err != nil {
		return false
	}

	// Must be inside home
	if !strings.HasPrefix(absPath, absHome) {
		return false
	}

	// Must not be home itself
	if absPath == absHome {
		return false
	}

	return true
}

// ExpandPath expands ~ and environment variables in a path.
func ExpandPath(path string) string {
	if strings.HasPrefix(path, "~") {
		usr, err := user.Current()
		if err == nil {
			path = usr.HomeDir + path[1:]
		}
	}
	return os.ExpandEnv(path)
}

// getHomeDir returns the user's home directory.
// It checks $HOME first for testability, then falls back to user.Current().
func getHomeDir() string {
	// $HOME is checked first so tests can override it with t.Setenv
	if home := os.Getenv("HOME"); home != "" {
		return home
	}
	usr, err := user.Current()
	if err == nil {
		return usr.HomeDir
	}
	return "/"
}

// HomeDir returns the user's home directory.
func HomeDir() string {
	return getHomeDir()
}
